controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,668 @@
+		return reconcile.Result{}, nil
+	}
+
+	// Track if the current Reconcile is doing a connect or disconnect.
+	didConnect := false
+	didDisconnect := false
+
+	requeueAfterDurations := []time.Duration{}
+
+	// Try to connect, if not connected.
+	connected := accessor.Connected(ctx)
+	if !connected {
+		lastConnectionCreationErrorTimestamp := accessor.GetLastConnectionCreationErrorTimestamp(ctx)
+
+		// Requeue, if connection creation failed within the ConnectionCreationRetryInterval.
+		if requeueAfter, requeue := shouldRequeue(time.Now(), lastConnectionCreationErrorTimestamp, accessor.config.ConnectionCreationRetryInterval); requeue {
+			log.V(6).Info(fmt.Sprintf("Requeuing after %s as connection creation already failed within the last %s",
+				requeueAfter.Truncate(time.Second/10), accessor.config.ConnectionCreationRetryInterval))
+			requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+		} else {
+			if err := accessor.Connect(ctx); err != nil {
+				// Requeue, if the connect failed.
+				log.V(6).Info(fmt.Sprintf("Requeuing after %s (connection creation failed)",
+					accessor.config.ConnectionCreationRetryInterval))
+				requeueAfterDurations = append(requeueAfterDurations, accessor.config.ConnectionCreationRetryInterval)
+			} else {
+				// Store that connect was done successfully.
+				didConnect = true
+				connected = true
+			}
+		}
+	}
+
+	// Run the health probe, if connected.
+	if connected {
+		lastProbeTimestamp := accessor.GetLastProbeTimestamp(ctx)
+
+		// Requeue, if health probe was already run within the HealthProbe.Interval.
+		if requeueAfter, requeue := shouldRequeue(time.Now(), lastProbeTimestamp, accessor.config.HealthProbe.Interval); requeue {
+			log.V(6).Info(fmt.Sprintf("Requeuing after %s as health probe was already run within the last %s",
+				requeueAfter.Truncate(time.Second/10), accessor.config.HealthProbe.Interval))
+			requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+		} else {
+			// Run the health probe
+			tooManyConsecutiveFailures, unauthorizedErrorOccurred := accessor.HealthCheck(ctx)
+			if tooManyConsecutiveFailures || unauthorizedErrorOccurred {
+				// Disconnect if the health probe failed (either with unauthorized or consecutive failures >= HealthProbe.FailureThreshold).
+				accessor.Disconnect(ctx)
+
+				// Store that disconnect was done.
+				didDisconnect = true
+				connected = false //nolint:ineffassign // connected is *currently* not used below, let's update it anyway
+			}
+			switch {
+			case unauthorizedErrorOccurred:
+				// Requeue for connection creation immediately.
+				// If we got an unauthorized error, it could be because the kubeconfig was rotated
+				// and in that case we want to immediately try to create the connection again.
+				log.V(6).Info("Requeuing immediately (disconnected after unauthorized error occurred)")
+				requeueAfterDurations = append(requeueAfterDurations, 1*time.Millisecond)
```

Comment:
What about going in reconcile backoff, so in case for any reason unauthorized persists we don't keep stuff in the working queue

Reply:
Rate-limiting today ensures we at most try to create a client every 30 second. If I'm starting to use exponential backoff the recovery time also goes up accordingly. Up until now I intentionally avoided this